### DISCUSSION TITLE
Which method does gyGeo uses to create final geometry

### DISCUSSION THREAD

**Discussion by Mohsin-Bashir** [mdolab/pygeo#233] [Q&A] [answered]
Hi i was doing Multidisciplinary Design Optimization using Dafoam https://dafoam.github.io/mydoc_get_started_faq.html#can-i-get-the-optimized-geometry-in-a-cad-format. After we have obtained optimized mesh, how does pyGeo use FFD volume with initial geometry to create final optimized CAD geometry.  
After studying the https://github.com/mdolab/pygeo/blob/main/pygeo/pyGeo.py and dvgeo.py files, what i understand is dvgeometry maps the ffd design variables information to initial geometry and then uses the same relationship to find the final geometry given the ffd design variables. Am i right? please guide me, if anyone have information about it.

**Comment by hntter** [comment]
hi, have you solved your problem?

**Reply by Mohsin-Bashir** [reply]
no

**Comment by lamkina** [comment, marked as answer]
Two main things I want to clarify:

1. pyGeo doesn't inherently store CAD geometry. Discrete mesh points are projected to a B-spline volume to invert the discrete cartesian coordinates to parametric coordinates of the volume.
2. As the B-spline volume (FFD) is deformed, we query the parametric coordinates to get the new discrete mesh coordinates.

In that sense, only two things are stored in pyGeo:
1. The surface mesh coordinates
2. The design variables that map to movement of the B-spline volume control points

If you want to get the optimal surface mesh, you supply the optimal design variables to DVGeometry and access the surface mesh coordinates or write them to a Tecplot file. You can also access the underlying deformed BSpline volume that is stored in DVGeometry as a pySpline object.

The lack of underlying CAD is a feature of pyGeo because most times we only work with discrete mesh representations of the geometry.

**Reply by WZDDZFJ** [reply]
I have a question about the gradient to ask you in DAFoam. On a two-dimensional airfoil, when it is not deformed, the gradient values of the FFD points symmetrical in the wingspan direction are very close. I can understand this. But why do they obtain different gradient values after synchronous deformation (that is, the result of calculating the gradient after the second time)? I can't understand this. They have exactly the same amount of deformation. Why are the gradient values after deformation significantly different ？ Is this also the reason for the boundary condition? 
 
![image](https://github.com/user-attachments/assets/ce235c09-842c-401e-9326-f8c2a666f8f9)
![image](https://github.com/user-attachments/assets/56446422-9eec-4de8-93f5-ab47de2d8edc)
